File: Java/cs225_java/src/cs22510_2014/GPSReader.java
package cs22510_2014;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Class representing a stream. Contains current location, 
 * current time and is responsible for reading from the file. 
 * Holds a boolean flag to indicate
 * whether the current satellite fix is good.
 * 
 * @author Evdzhan Mustafa enm3@aber.ac.uk
 * 
 */
public class GPSReader {

	/**
	 * Nested class to represent a location and the time.
	 * 
	 * @author Evdzhan Mustafa enm3@aber.ac.uk
	 * 
	 */
	protected class Location {
		protected double lat; // latitude
		protected double lng; // longitude
		protected Date date;

	}

	/* A various static integers for the return value of read function. */
	protected final static int _EOF = -1;
	protected final static int SKIP_LINE = 1;
	protected final static int GPS_TIME = 2;
	protected final static int SATELITE = 3;

	/* Strings to detect the lines we need. */
	protected final static String GSV = "GSV";
	protected final static String RMC = "RMC";

	protected final static double MIL = 1000000.0; // used in rounding

	protected Location currLoc;
	protected Date currTime;
	protected boolean satellitesOK;

	private BufferedReader br;

	/**
	 * @param fileName  The file to be opened
	 */

	public GPSReader(String fileName) {
		try {
			br = new BufferedReader(new FileReader(fileName));
		} catch (IOException e) {
			e.printStackTrace();
		}

		// assume satellite fix is bad at the beginning
		satellitesOK = false; 

	}

	/**
	 * Reads a sentence from the file.
	 * 
	 * @return integer corresponding to the type of sentence read
	 */
	protected int read() {
		String data = "";
		try {

			data = br.readLine();
			if (data == null)
				return _EOF; // end of file reached

			if (data.equals("")){
				return SKIP_LINE; // empty line
				} 
			
            // see what type of sentence was  read
			String type = data.substring(3, 6); 
												

			if (type.equals(GSV)) {

				// get the number of gsv sentences
				int linesNum = Integer.parseInt(data.substring(7, 8));

				// array to insert the gsv sentences
				String[] sateliteData = new String[linesNum];

				// place the already read gsv line to the array
				sateliteData[0] = data;
				
				
				// store each gsv sentence in the array
				for (int i = 1; i < linesNum; i++) { 
														 
					sateliteData[i] = br.readLine();
				}

				// this will update the satellitesOK boolean
				processGSV(sateliteData); 
											 
				// satellite info obtained
				return SATELITE; 

			} else if (type.equals(RMC)) {

				processRMC(data);

				return GPS_TIME; // location and time updated

			} else {

				return SKIP_LINE; // skip any other line

			}

		} catch (Exception e) {
			e.printStackTrace();
			return _EOF; // if error better exit
		}

	}

	/**
	 * Works on the data given by GSV sentence(s)
	 * 
	 * @param gsvData  array containing all the gsv sentences
	 */
	private void processGSV(String[] gsvData) {

		int satNum = Integer.parseInt(gsvData[0].substring(11, 13));

		int count = 0;

		
		// loop through the gsv sentences
		for (int i = 0; i < gsvData.length; i++) { 
												 

			// get the index of the asterix
			int asterixGetter = gsvData[i].indexOf('*'); 
															  

			// split the current sentence to tokens
			String[] tokens = gsvData[i].substring(14, asterixGetter)
					.split(",");

			// loop through the SNR values
			for (int o = 3; o < 16 && satNum > 0 && 
					o < tokens.length; o += 4, satNum--) {

				// check if snr is good  
				if (!tokens[o].equals("") &&
						Integer.parseInt(tokens[o]) >= 30) {
					count++;
					if (count == 3) {

						// satellites are ok
						this.satellitesOK = true; 
						
						// no point looking for further SNR sentences 
						return; 							 
					}
				}
			}
		}
		 // if this line is reached, satellites are NOT ok
		this.satellitesOK = false;
	}

	/**
	 * Processes a RMC sentence. Gets the date time and the new coordinates.
	 * 
	 * @param rmcData
	 */
	private void processRMC(String rmcData) {

		String[] thedata = rmcData.split(",");
		
		// drop the milliseconds
		String thetime = thedata[1].substring(0, 6); 
		String thedate = thedata[9];

		try {
			currTime = new SimpleDateFormat("HHmmssddMMyy", Locale.ENGLISH)
					.parse(thetime + thedate);

		} catch (ParseException e) {
			e.printStackTrace();
		}

		// turn the degrees to decimal and place it to current loc
		this.currLoc = degreesToDecimal(thedata[3], thedata[5]);

		if (thedata[4].equals("S")) // if in the south hemisphere
			this.currLoc.lat *= -1;

		if (thedata[6].equals("W")) // if in the west hemisphere
			this.currLoc.lng *= -1;

		this.currLoc.date = currTime;

	}

	/**
	 * Converts degrees and minute representation 
	 * of latitude and longitude, to a decimal form.
	 * 
	 * @param latitude  string holding latitude
	 * @param longitude string holding longitude
	 * @return the location with decimal coordinates
	 */
	private Location degreesToDecimal(String latitude, String longitude) {

		Location loc = new Location();

		// parse the value first
		double lat = Double.parseDouble(latitude);

		// truncate to get the degrees
		int lat_degrees = (int) lat / 100;

		// transform the minutes to decimal
		double lat_minutes = (lat - lat_degrees * 100) / 60.0;

		
		
		// repeat for longitude
		double lng = Double.parseDouble(longitude);
		int lng_degrees = (int) lng / 100;
		double lng_minutes = (lng - lng_degrees * 100) / 60.0;

		
		// do some rounding at place the new data to the location
		loc.lat = Math.round((lat_minutes + lat_degrees) * MIL) / MIL;
		loc.lng = Math.round((lng_minutes + lng_degrees) * MIL) / MIL;

		return loc;
	}
}

File: old/CC++/cs225_c++/GPSReader.cpp
/* 
 * File:   GPSReader.cpp
 * Author: evdjoint
 * 
 * Created on 25 March 2014, 16:39
 */
#include <string>
#include <sstream>
#include <vector>
#include <iostream>
#include <fstream>
#include <math.h>
#include "Location.h"
#include "GPSReader.h"
using namespace std;

GPSReader::GPSReader(const std::string& file) {
    reader.open(file.c_str());
}

/* Reads a line from the stream, 
 * returns integer based on what was read.*/
int GPSReader::read() {

    string line_read; // string to place the line

    if (!getline(reader, line_read)) {

        return _EOF; // stream ended

    } else if (line_read.empty() ||
            line_read.compare("\r") == 0) {

        return SKIP_LINE; // empty line, skip it

    } else if (line_read.substr(3, 3) == "RMC") {

        // process the new coordinates and time
        process_rmc(line_read);

        return GPS_TIME;


    } else if (line_read.substr(3, 3) == "GSV") {

        // see how many gsv sentences are there
        stringstream strstream(line_read.substr(7, 1));

        int num_lines;
        strstream >> num_lines;

        //create container array 
        //to put the gsv sentences
        string lines[num_lines];

        // the first sentences is already in line_read
        lines[0] = line_read;

        int i = 1;

        // put  the sentences in the container
        for (i; i < num_lines; i++) {
            getline(reader, line_read);
            lines[i] = line_read;
        }

        // process the sentences
        process_gsv(lines, num_lines);
        return SATELITE;

    } else {

        // skip if anything else occurred 
        return SKIP_LINE; 
    }

}

/* Read the gsv sentences in the data array.
 * Determines whether the satellite fix is ok. */
void GPSReader::process_gsv(string data[], int num_sentences) {

    int count = 0;
    int num_satelites;


    stringstream strstream(data[0].substr(11, 2));

    // parse the number of satellites 
    strstream >> num_satelites;
    int i = 0;

    // loop through the sentences 
    for (i = 0; i < num_sentences; ++i) {

        stringstream asterix_getter(data[i]);

        // get the sentence up to the asterix
        getline(asterix_getter, data[i], '*');

        // vector to place the tokens
        vector<string> tokens; 

        // treat the sentence as string stream
        stringstream ss(data[i]); 
        string current;

        // split the current sentence to tokens
        while (getline(ss, current, ',')) {
            tokens.push_back(current);
        }

        int o = 0;
        //loop through the satellites 
        for (o = 7; o < tokens.size() && !tokens.at(o).empty()
                && num_satelites > 0; o += 4, num_satelites--) {

            int value;
            stringstream satelitegetter(tokens.at(o));
            
            // parse the SNR
            satelitegetter >> value; 

            // if SNR good increment count
            if (value >= 30) { 
                count++;

                // at least 3 satellites had SNR of 30 or more
                if (count == 3) {
                    satelitesOK = true;
                    return; //  we are done here
                }
            }
        }
    }
    //if the code reaches here, no 3 satellites were  good
    satelitesOK = false;
    return;


}

void GPSReader::process_rmc(string& data) {

    stringstream asterix_getter(data);

    // get the line up to the asterix 
    getline(asterix_getter, data, '*');


    //vector to store the tokens
    vector<string> tokens;

    stringstream ss(data);
    string current;


    // split the string to tokens
    while (getline(ss, current, ',')) {
        tokens.push_back(current);
    }


    // get the time
    stringstream time_getter(tokens.at(1));

    string time;

    getline(time_getter, time, '.');


    // concatenate the time with the date
    time += tokens.at(9);

    // take the time and transform it into struct tm time
    strptime(time.c_str(), "%H%M%S%d%m%y", &this->loc.the_time);

    
    // transform the degrees
    degrees_to_decimal(tokens.at(3), tokens.at(5)); 

    // if the coordinate is in the southern hemisphere
    if (tokens.at(4) == "S") {
        loc.latitude *= -1;

    }
    
    // if the coordinate is in the western hemisphere
    if (tokens.at(6) == "W") {
        loc.longitude *= -1;
    }

}

void GPSReader::degrees_to_decimal(string lat, string lng) {

    stringstream convertlat(lat);
    double lat_;
    convertlat >>lat_; // parse the latitude
    int lat_degrees = (int) lat_ / 100; // truncate the degrees

    //turn minutes to decimal
    double lat_minutes = (lat_ - lat_degrees * 100) / 60.0;


    //now repeat for the longitude
    stringstream convertlng(lng);
    double lng_;
    convertlng >> lng_;
    int lng_degrees = (int) lng_ / 100;
    double lng_minutes = (lng_ - lng_degrees * 100) / 60.0;



    //round both values and put them to the location
    loc.latitude = round((lat_degrees + lat_minutes) * MIL) / MIL;
    loc.longitude = round((lng_degrees + lng_minutes) * MIL) / MIL;


}

GPSReader::GPSReader(const GPSReader& orig) {
}

GPSReader::~GPSReader() {
    reader.close();
}

